src/java.base/share/classes/java/io/FileDescriptor.java

```diff
@@ -297,6 +368,24 @@ synchronized void close() throws IOException {
         close0();
     }
 
+    private synchronized void beforeCheckpoint() throws CheckpointOpenFileException {
+        if (valid()) {
+            JDKContext ctx = jdk.internal.crac.Core.getJDKContext();
+            if (ctx.claimFdWeak(this, this)) {
+                String msg = "FileDescriptor " + this.fd + " left open. ";
+                if (!JDKContext.Properties.COLLECT_FD_STACKTRACES) {
+                    msg += JDKContext.COLLECT_FD_STACKTRACES_HINT;
+                }
+                throw new CheckpointOpenFileException(msg, resource.stackTraceHolder);
```

Comment:
Now, the message looks like:

```
	Suppressed: jdk.crac.impl.CheckpointOpenFileException: FileDescriptor 323 left open.
		at java.base/java.io.FileDescriptor.beforeCheckpoint(FileDescriptor.java:379)
		at java.base/java.io.FileDescriptor$Resource.beforeCheckpoint(FileDescriptor.java:80)
		at java.base/jdk.crac.impl.AbstractContextImpl.runBeforeCheckpoint(AbstractContextImpl.java:114)
		at java.base/jdk.crac.impl.AbstractContextImpl.beforeCheckpoint(AbstractContextImpl.java:83)
		at java.base/jdk.internal.crac.JDKContext.beforeCheckpoint(JDKContext.java:85)
		at java.base/jdk.crac.impl.AbstractContextImpl.runBeforeCheckpoint(AbstractContextImpl.java:114)
		at java.base/jdk.crac.impl.AbstractContextImpl.beforeCheckpoint(AbstractContextImpl.java:83)
		at java.base/jdk.crac.Core.checkpointRestore1(Core.java:121)
		... 2 more
	Caused by: java.lang.Exception: This file descriptor was created here
		at java.base/java.io.FileDescriptor$Resource.<init>(FileDescriptor.java:71)
		at java.base/java.io.FileDescriptor.<init>(FileDescriptor.java:100)
		at java.base/sun.nio.ch.IOUtil.newFD(IOUtil.java:544)
		at java.base/sun.nio.ch.Net.socket(Net.java:524)
		at java.base/sun.nio.ch.SocketChannelImpl.<init>(SocketChannelImpl.java:146)
		at java.base/sun.nio.ch.SocketChannelImpl.<init>(SocketChannelImpl.java:129)
		at java.base/sun.nio.ch.SelectorProviderImpl.openSocketChannel(SelectorProviderImpl.java:77)
		at java.base/java.nio.channels.SocketChannel.open(SocketChannel.java:192)
```

While in the past the FD (which is actually a socket, not a file) was reported by the VM like:

```
	Suppressed: jdk.crac.impl.CheckpointOpenSocketException: tcp localAddr 0.0.0.0 localPort 8080 remoteAddr 0.0.0.0 remotePort 0
		at java.base/jdk.crac.Core.translateJVMExceptions(Core.java:92)
		at java.base/jdk.crac.Core.checkpointRestore1(Core.java:157)
```

We miss the fact this is a socket now, as well as details (although the stack trace is very useful!).

We can ask the VM for the FD type and details. To so we'll report report the details as we used to, by reusing the existing code providing the details, that should not be very hard.